File: packages/windows-dashboard/Source/DisplayGeometry.cs
using System.Windows;

namespace ClassroomWidgets;

public static class DisplayGeometry
{
    public static Rect? AspectFit(Size content, Rect bounds)
    {
        if (!Positive(content) || !Positive(bounds.Size) || !Finite(bounds)) return null;
        var scale = Math.Min(bounds.Width / content.Width, bounds.Height / content.Height);
        var width = content.Width * scale;
        var height = content.Height * scale;
        return new Rect(bounds.X + (bounds.Width - width) / 2, bounds.Y + (bounds.Height - height) / 2, width, height);
    }

    public static Point? MapPreviewPointToSource(Point point, Rect imageRect, Rect sourceBounds)
    {
        if (!Finite(point) || !Positive(imageRect.Size) || !Finite(imageRect) || !Positive(sourceBounds.Size) || !Finite(sourceBounds)) return null;
        var u = (point.X - imageRect.X) / imageRect.Width;
        var v = (point.Y - imageRect.Y) / imageRect.Height;
        if (u < 0 || u >= 1 || v < 0 || v >= 1) return null;
        var x = Math.Min(sourceBounds.X + u * sourceBounds.Width, sourceBounds.Right - 1);
        var y = Math.Min(sourceBounds.Y + v * sourceBounds.Height, sourceBounds.Bottom - 1);
        return double.IsFinite(x) && double.IsFinite(y) ? new Point(x, y) : null;
    }

    public static Size AspectNormalizedWindowSize(
        double aspect,
        Size proposedPreview,
        double chromeHeight,
        Size minimumPreview,
        Size maximum)
    {
        var chrome = double.IsFinite(chromeHeight) ? Math.Max(chromeHeight, 0) : 0;
        var fallback = new Size(proposedPreview.Width, proposedPreview.Height + chrome);
        if (!double.IsFinite(aspect) || aspect <= 0 || !Positive(proposedPreview)) return fallback;

        var minimum = new Size(
            Math.Max(double.IsFinite(minimumPreview.Width) ? minimumPreview.Width : 1, 1),
            Math.Max(double.IsFinite(minimumPreview.Height) ? minimumPreview.Height : 1, 1));
        var scale = Math.Min(proposedPreview.Width / aspect, proposedPreview.Height);
        var width = aspect * scale;
        var height = scale;
        if (width < minimum.Width || height < minimum.Height)
        {
            var minimumScale = Math.Max(minimum.Width / aspect, minimum.Height);
            width = aspect * minimumScale;
            height = minimumScale;
        }
        if (Positive(maximum))
        {
            var maximumPreviewHeight = Math.Max(maximum.Height - chrome, 1);
            var screenScale = Math.Min(1, Math.Min(maximum.Width / width, maximumPreviewHeight / height));
            if (screenScale < 1)
            {
                var scaledWidth = width * screenScale;
                var scaledHeight = height * screenScale;
                if (scaledWidth >= minimum.Width && scaledHeight >= minimum.Height)
                {
                    width = scaledWidth;
                    height = scaledHeight;
                }
                else
                {
                    width = Math.Max(minimum.Width, Math.Min(width, maximum.Width));
                    height = Math.Max(minimum.Height, Math.Min(height, maximumPreviewHeight));
                }
            }
        }
        return new Size(width, height + chrome);
    }

    public static bool Intersects(Rect a, Rect b)
        => Finite(a) && Finite(b) && Math.Min(a.Right, b.Right) > Math.Max(a.Left, b.Left)
            && Math.Min(a.Bottom, b.Bottom) > Math.Max(a.Top, b.Top);

    public static Rect Clamp(Rect rect, Rect bounds)
    {
        if (!Finite(rect) || !Finite(bounds)) return rect;
        var width = Math.Min(Math.Max(rect.Width, 0), Math.Max(bounds.Width, 0));
        var height = Math.Min(Math.Max(rect.Height, 0), Math.Max(bounds.Height, 0));
        var x = Math.Min(Math.Max(rect.X, bounds.X), bounds.Right - width);
        var y = Math.Min(Math.Max(rect.Y, bounds.Y), bounds.Bottom - height);
        return new Rect(x, y, width, height);
    }

    public static string DescribePosition(Rect bounds)
    {
        if (bounds.X < 0) return "left";
        if (bounds.Y < 0) return "above";
        if (bounds.X > 0) return "right";
        if (bounds.Y > 0) return "below";
        return "main";
    }

    private static bool Positive(Size size)
        => double.IsFinite(size.Width) && double.IsFinite(size.Height) && size.Width > 0 && size.Height > 0;

    private static bool Finite(Point point) => double.IsFinite(point.X) && double.IsFinite(point.Y);

    private static bool Finite(Rect rect)
        => double.IsFinite(rect.X) && double.IsFinite(rect.Y) && double.IsFinite(rect.Width) && double.IsFinite(rect.Height);
}

File: packages/linux-dashboard/src/main/displayGeometry.ts
export interface Rect {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface Size {
  width: number;
  height: number;
}

export interface Point {
  x: number;
  y: number;
}

function positiveFiniteSize(size: Size): boolean {
  return Number.isFinite(size.width) && Number.isFinite(size.height) && size.width > 0 && size.height > 0;
}

function finiteRect(rect: Rect): boolean {
  return [rect.x, rect.y, rect.width, rect.height].every(Number.isFinite);
}

export function aspectFit(contentSize: Size, bounds: Rect): Rect | null {
  if (!positiveFiniteSize(contentSize) || !finiteRect(bounds) || bounds.width <= 0 || bounds.height <= 0) return null;
  const scale = Math.min(bounds.width / contentSize.width, bounds.height / contentSize.height);
  const width = contentSize.width * scale;
  const height = contentSize.height * scale;
  return {
    x: bounds.x + (bounds.width - width) / 2,
    y: bounds.y + (bounds.height - height) / 2,
    width,
    height,
  };
}

export function mapPreviewPointToSource(point: Point, imageRect: Rect, sourceBounds: Rect): Point | null {
  if (![point.x, point.y].every(Number.isFinite)
    || !finiteRect(imageRect) || imageRect.width <= 0 || imageRect.height <= 0
    || !finiteRect(sourceBounds) || sourceBounds.width <= 0 || sourceBounds.height <= 0) return null;
  const u = (point.x - imageRect.x) / imageRect.width;
  const v = (point.y - imageRect.y) / imageRect.height;
  if (!(u >= 0 && u < 1 && v >= 0 && v < 1)) return null;
  const x = Math.min(sourceBounds.x + u * sourceBounds.width, sourceBounds.x + sourceBounds.width - 1);
  const y = Math.min(sourceBounds.y + v * sourceBounds.height, sourceBounds.y + sourceBounds.height - 1);
  return Number.isFinite(x) && Number.isFinite(y) ? { x, y } : null;
}

export function aspectNormalizedWindowSize(
  aspect: number,
  proposedPreviewSize: Size,
  chromeHeight: number,
  minimumPreviewSize: Size,
  maximumSize: Size,
): Size {
  const chrome = Number.isFinite(chromeHeight) ? Math.max(chromeHeight, 0) : 0;
  const fallback = { width: proposedPreviewSize.width, height: proposedPreviewSize.height + chrome };
  if (!Number.isFinite(aspect) || aspect <= 0 || !positiveFiniteSize(proposedPreviewSize)) return fallback;

  const minimum = {
    width: Math.max(Number.isFinite(minimumPreviewSize.width) ? minimumPreviewSize.width : 1, 1),
    height: Math.max(Number.isFinite(minimumPreviewSize.height) ? minimumPreviewSize.height : 1, 1),
  };
  const scale = Math.min(proposedPreviewSize.width / aspect, proposedPreviewSize.height);
  let width = aspect * scale;
  let height = scale;
  if (width < minimum.width || height < minimum.height) {
    const minimumScale = Math.max(minimum.width / aspect, minimum.height);
    width = aspect * minimumScale;
    height = minimumScale;
  }
  if (positiveFiniteSize(maximumSize)) {
    const maximumPreviewHeight = Math.max(maximumSize.height - chrome, 1);
    const screenScale = Math.min(1, maximumSize.width / width, maximumPreviewHeight / height);
    if (screenScale < 1) {
      const scaledWidth = width * screenScale;
      const scaledHeight = height * screenScale;
      if (scaledWidth >= minimum.width && scaledHeight >= minimum.height) {
        width = scaledWidth;
        height = scaledHeight;
      } else {
        width = Math.max(minimum.width, Math.min(width, maximumSize.width));
        height = Math.max(minimum.height, Math.min(height, maximumPreviewHeight));
      }
    }
  }
  return { width, height: height + chrome };
}

export function rectsIntersect(a: Rect, b: Rect): boolean {
  return finiteRect(a) && finiteRect(b)
    && Math.min(a.x + a.width, b.x + b.width) > Math.max(a.x, b.x)
    && Math.min(a.y + a.height, b.y + b.height) > Math.max(a.y, b.y);
}

export function clampRect(rect: Rect, bounds: Rect): Rect {
  if (!finiteRect(rect) || !finiteRect(bounds)) return rect;
  const width = Math.min(Math.max(rect.width, 0), Math.max(bounds.width, 0));
  const height = Math.min(Math.max(rect.height, 0), Math.max(bounds.height, 0));
  return {
    x: Math.min(Math.max(rect.x, bounds.x), bounds.x + bounds.width - width),
    y: Math.min(Math.max(rect.y, bounds.y), bounds.y + bounds.height - height),
    width,
    height,
  };
}

export function intersectionArea(a: Rect, b: Rect): number {
  if (!rectsIntersect(a, b)) return 0;
  return (Math.min(a.x + a.width, b.x + b.width) - Math.max(a.x, b.x))
    * (Math.min(a.y + a.height, b.y + b.height) - Math.max(a.y, b.y));
}

export function describePosition(bounds: Rect): string {
  if (bounds.x < 0) return 'left';
  if (bounds.y < 0) return 'above';
  if (bounds.x > 0) return 'right';
  if (bounds.y > 0) return 'below';
  return 'main';
}
